dali/kernels/imgproc/warp/warp_setup.cuh

```diff
@@ -0,0 +1,232 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_IMGPROC_WARP_WARP_SETUP_CUH_
+#define DALI_KERNELS_IMGPROC_WARP_WARP_SETUP_CUH_
+
+#include <vector>
+#include <utility>
+#include "dali/kernels/kernel.h"
+#include "dali/core/geom/vec.h"
+
+namespace dali {
+namespace kernels {
+namespace warp {
+
+template <int ndim>
+struct SampleDesc {
+  const void *__restrict__ input;
+  void *__restrict__ output;
+  ivec<ndim> out_size, out_strides, in_size, in_strides;
+  int channels;
+};
+
+template <int ndim>
+struct BlockDesc {
+  int sample_idx;
+  ivec<ndim> start, end;
+};
+
+template <int ndim>
+class WarpSetup {
+  static_assert(ndim == 2 || ndim == 3,
+    "Warping is defined only for 2D and 3D data with interleaved channels");
+
+ public:
+  static constexpr int tensor_dim = ndim+1;
+  using SampleDesc = warp::SampleDesc<ndim>;
+  using BlockDesc = warp::BlockDesc<ndim>;
+
+  KernelRequirements Setup(const TensorListShape<tensor_dim> &output_shape,
+                           bool force_variable_size = false) {
+    ScratchpadEstimator se;
+    is_uniform_ = !force_variable_size && is_uniform(output_shape);
+    if (is_uniform_)
+      UniformSizeSetup(se, output_shape);
+    else
+      VariableSizeSetup(se, output_shape);
+
+    KernelRequirements req = {};
+    req.output_shapes = { std::move(output_shape) };
+    req.scratch_sizes = se.sizes;
+    return req;
+  }
+
+  TensorListShape<tensor_dim> GetOutputShape(
+      const TensorListShape<tensor_dim> &in_shape,
+      span<const TensorShape<ndim>> output_sizes) {
+    TensorListShape<tensor_dim> shape;
+    shape.resize(in_shape.num_samples(), tensor_dim);
+    for (int i = 0; i < in_shape.num_samples(); i++) {
```

Comment:
Does it make any sense to extract some common part of `GetOutputShape` and `ValidateOutputShape`.